{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0e3ca29e-12ef-4153-8fa2-96b279a0d3df",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f067f1b-f9c8-45ad-b180-8ea6767b4278",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard Library imports\n",
    "import os\n",
    "import pathlib\n",
    "\n",
    "# External Imports\n",
    "import cobra\n",
    "from cobra.flux_analysis import single_gene_deletion\n",
    "import metworkpy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Local Imports"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f20b813-34af-4085-bdf1-a09844816aeb",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a368da38-3c86-48e1-8017-532672075431",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.Configuration.solver = \"cplex\"\n",
    "## Local Developement\n",
    "# BASE_PATH = pathlib.Path(\"..\")\n",
    "## HPC\n",
    "HOME = os.getenv(\"HOME\")\n",
    "BASE_PATH = pathlib.Path(HOME) / \"projects\" / \"KiMAT\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f80f06ea-7bc5-4c0f-870c-b44e60708da6",
   "metadata": {},
   "source": [
    "## Data Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13feb7a8-f1cf-4da3-a875-380018ecb17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in gene expression\n",
    "gene_expression = pd.read_csv(BASE_PATH / 'data' / 'gene-expression' / 'kinase_RNA_seq.csv', index_col=0).drop( \"GrowthPhase\", axis=1)\n",
    "# Convert the RPKM measurements into TPM\n",
    "gene_expression = metworkpy.utils.rpkm_to_tpm(gene_expression)\n",
    "# Read in quantile normalized gene expression\n",
    "normalized_expression = pd.read_csv(BASE_PATH / \"data\" / \"quant_normed\" / \"xprs_norm.tsv\", sep=\"\\t\", index_col=0).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc0d5e4-0d98-4179-9ae3-a447b8064942",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate the expression data into biological replicates\n",
    "expr_dict = {\n",
    "    \"WT\":None,\n",
    "    \"PknB_KD\":None,\n",
    "    \"PknB_IND\":None,\n",
    "    \"PknD_KO\":None,\n",
    "    \"PknD_IND\":None,\n",
    "    \"PknE_KO\":None,\n",
    "    \"PknE_IND\":None,\n",
    "    \"PknF_KO\":None,\n",
    "    \"PknF_IND\":None,\n",
    "    \"PknG_KO\":None,\n",
    "    \"PknG_IND\":None,\n",
    "    \"PknH_KO\":None,\n",
    "    \"PknH_IND\":None,\n",
    "    \"PknJ_KO\":None,\n",
    "    \"PknJ_IND\":None,\n",
    "    \"PknK_KO\":None,\n",
    "    \"PknK_IND\":None,\n",
    "    \"PknL_KO\":None,\n",
    "    \"PknL_IND\":None,\n",
    "}\n",
    "norm_expr_dict = {\n",
    "    \"WT\":None,\n",
    "    \"PknB_KD\":None,\n",
    "    \"PknB_IND\":None,\n",
    "    \"PknD_KO\":None,\n",
    "    \"PknD_IND\":None,\n",
    "    \"PknE_KO\":None,\n",
    "    \"PknE_IND\":None,\n",
    "    \"PknF_KO\":None,\n",
    "    \"PknF_IND\":None,\n",
    "    \"PknG_KO\":None,\n",
    "    \"PknG_IND\":None,\n",
    "    \"PknH_KO\":None,\n",
    "    \"PknH_IND\":None,\n",
    "    \"PknJ_KO\":None,\n",
    "    \"PknJ_IND\":None,\n",
    "    \"PknK_KO\":None,\n",
    "    \"PknK_IND\":None,\n",
    "    \"PknL_KO\":None,\n",
    "    \"PknL_IND\":None,\n",
    "}\n",
    "\n",
    "\n",
    "for condition in expr_dict:\n",
    "    expr_dict[condition] = gene_expression.loc[gene_expression.index.str.startswith(condition)]\n",
    "    norm_expr_dict[condition] = normalized_expression.loc[normalized_expression.index.str.startswith(condition)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e32e0c2-f7ca-4989-9ece-94cc09bda386",
   "metadata": {},
   "source": [
    "## Read in Base Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b0e4ba-0fc5-4c70-9a4e-4afce7ff8a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "iek1011 = metworkpy.read_model(BASE_PATH / \"data\" / \"Models\" / \"iEK1011_m7H10_media.json\")\n",
    "for method in [\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"]:\n",
    "    metworkpy.write_model(iek1011, \n",
    "                          BASE_PATH / \"results\" / \"diff_expr_models\" / method / \"iek1011_WT_model.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d120951-2359-4504-9cc1-82ee48cad2e5",
   "metadata": {},
   "source": [
    "## Convert Gene Expression into Trinarized Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5db11b6-e32a-45c8-9164-ca55ce7614cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "rxn_weight_dict = {\n",
    "    \"WT\":None,\n",
    "    \"PknB_KD\":None,\n",
    "    \"PknB_IND\":None,\n",
    "    \"PknD_KO\":None,\n",
    "    \"PknD_IND\":None,\n",
    "    \"PknE_KO\":None,\n",
    "    \"PknE_IND\":None,\n",
    "    \"PknF_KO\":None,\n",
    "    \"PknF_IND\":None,\n",
    "    \"PknG_KO\":None,\n",
    "    \"PknG_IND\":None,\n",
    "    \"PknH_KO\":None,\n",
    "    \"PknH_IND\":None,\n",
    "    \"PknJ_KO\":None,\n",
    "    \"PknJ_IND\":None,\n",
    "    \"PknK_KO\":None,\n",
    "    \"PknK_IND\":None,\n",
    "    \"PknL_KO\":None,\n",
    "    \"PknL_IND\":None,\n",
    "}\n",
    "norm_rxn_weight_dict = {\n",
    "    \"WT\":None,\n",
    "    \"PknB_KD\":None,\n",
    "    \"PknB_IND\":None,\n",
    "    \"PknD_KO\":None,\n",
    "    \"PknD_IND\":None,\n",
    "    \"PknE_KO\":None,\n",
    "    \"PknE_IND\":None,\n",
    "    \"PknF_KO\":None,\n",
    "    \"PknF_IND\":None,\n",
    "    \"PknG_KO\":None,\n",
    "    \"PknG_IND\":None,\n",
    "    \"PknH_KO\":None,\n",
    "    \"PknH_IND\":None,\n",
    "    \"PknJ_KO\":None,\n",
    "    \"PknJ_IND\":None,\n",
    "    \"PknK_KO\":None,\n",
    "    \"PknK_IND\":None,\n",
    "    \"PknL_KO\":None,\n",
    "    \"PknL_IND\":None,\n",
    "}\n",
    "\n",
    "diff_rxn_weight_dict = {\n",
    "    \"PknB_KD\":None,\n",
    "    \"PknB_IND\":None,\n",
    "    \"PknD_KO\":None,\n",
    "    \"PknD_IND\":None,\n",
    "    \"PknE_KO\":None,\n",
    "    \"PknE_IND\":None,\n",
    "    \"PknF_KO\":None,\n",
    "    \"PknF_IND\":None,\n",
    "    \"PknG_KO\":None,\n",
    "    \"PknG_IND\":None,\n",
    "    \"PknH_KO\":None,\n",
    "    \"PknH_IND\":None,\n",
    "    \"PknJ_KO\":None,\n",
    "    \"PknJ_IND\":None,\n",
    "    \"PknK_KO\":None,\n",
    "    \"PknK_IND\":None,\n",
    "    \"PknL_KO\":None,\n",
    "    \"PknL_IND\":None,\n",
    "}\n",
    "\n",
    "wt_expression = gene_expression.loc[gene_expression.index.str.startswith(\"WT\")].median(axis=0)+1e-10\n",
    "\n",
    "for condition in rxn_weight_dict:\n",
    "    p = BASE_PATH / 'cache' / 'rxn_weights' / f'{condition}_rxn_weights.json'\n",
    "    norm_p = BASE_PATH / 'cache' /'norm_rxn_weights' / f'{condition}_rxn_weights.json'\n",
    "    if p.exists():\n",
    "        rxn_weight_dict[condition] = pd.read_json(p, typ = \"series\")\n",
    "    else: \n",
    "        p.parent.mkdir(parents=True, exist_ok = True)\n",
    "        gene_weights = metworkpy.utils.expr_to_weights(expression = expr_dict[condition],\n",
    "                                                      quantile = 0.15, \n",
    "                                                      subset = iek1011.genes.list_attr(\"id\"),\n",
    "                                                      sample_axis=0)\n",
    "        rxn_weight_dict[condition] = metworkpy.parse.gene_to_rxn_weights(iek1011, gene_weights)\n",
    "        rxn_weight_dict[condition].to_json(p)\n",
    "    if norm_p.exists():\n",
    "        norm_rxn_weight_dict[condition] = pd.read_json(norm_p, typ = \"series\")\n",
    "    else: \n",
    "        norm_p.parent.mkdir(parents=True, exist_ok = True)\n",
    "        norm_gene_weights = metworkpy.utils.expr_to_weights(expression = norm_expr_dict[condition],\n",
    "                                                      quantile = 0.15, \n",
    "                                                      subset = iek1011.genes.list_attr(\"id\"),\n",
    "                                                      sample_axis=0)\n",
    "        norm_rxn_weight_dict[condition] = metworkpy.parse.gene_to_rxn_weights(iek1011, norm_gene_weights)\n",
    "        norm_rxn_weight_dict[condition].to_json(norm_p)\n",
    "    if condition != \"WT\":\n",
    "        diff_p = BASE_PATH / \"cache\" / \"diff_rxn_weights\" / f\"{condition}_rxn_weights.json\"\n",
    "        if diff_p.exists():\n",
    "            diff_rxn_weight_dict[condition] = pd.read_json(diff_p, typ=\"series\")\n",
    "        else:\n",
    "            diff_p.parent.mkdir(parents=True, exist_ok = True)\n",
    "            log2fc = np.log2(gene_expression.loc[gene_expression.index.str.startswith(condition)].median(axis=0).div(wt_expression)+1e-10)\n",
    "            diff_gene_weights = log2fc.apply(lambda x: -1 if x<=-1 else (1 if x>=1 else 0))\n",
    "            diff_rxn_weight_dict[condition] = metworkpy.parse.gene_to_rxn_weights(iek1011, diff_gene_weights)\n",
    "            diff_rxn_weight_dict[condition].to_json(diff_p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92f07d59-0d50-49dc-8eac-de989d459909",
   "metadata": {},
   "source": [
    "## IMAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624c1ffb-a7dc-4ace-9b54-031954d77eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "biomass_growth_df = pd.DataFrame(np.NaN, index = rxn_weight_dict.keys(), \n",
    "                                 columns=[\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"],\n",
    "                                dtype=\"float\")\n",
    "norm_biomass_growth_df = pd.DataFrame(np.NaN, index = norm_rxn_weight_dict.keys(), \n",
    "                                 columns=[\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"],\n",
    "                                dtype=\"float\")\n",
    "diff_biomass_growth_df = pd.DataFrame(np.NaN, index = norm_rxn_weight_dict.keys(), \n",
    "                                     columns = [\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"],\n",
    "                                     dtype=\"float\")\n",
    "wt_objective = iek1011.slim_optimize()\n",
    "diff_biomass_growth_df.loc['WT'] = wt_objective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08d2190f-99f7-4df1-bed8-9c1c7cee68fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def imat_per_method(method, weights_dict, results_path, biomass_growth_df, skip_list=None):\n",
    "    if skip_list is None:\n",
    "        skip_list = []\n",
    "    print(f\"METHOD: {method}\")\n",
    "    print(\"**********************\")\n",
    "    for condition in weights_dict:\n",
    "        print(f\"\\tCondition: {condition}\")\n",
    "        print(\"\\t--------------\")\n",
    "        if condition in skip_list:\n",
    "            print(\"SKIPPED\")\n",
    "            continue\n",
    "        p = results_path / method / f\"iek1011_{condition}_model.json\"\n",
    "        if p.exists():\n",
    "            model = metworkpy.read_model(str(p))\n",
    "        else:\n",
    "            p.parent.mkdir(parents=True, exist_ok = True)\n",
    "            try:\n",
    "                model = metworkpy.imat.generate_model(model = iek1011.copy(), \n",
    "                                                      rxn_weights=weights_dict[condition],\n",
    "                                                     method = method)\n",
    "                metworkpy.write_model(model, model_path=p)\n",
    "            except Exception as exp:\n",
    "                continue\n",
    "        try:\n",
    "            biomass_growth_df.loc[condition, method] = model.slim_optimize()\n",
    "        except Exception as exp:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06660665-c600-4e8a-9215-759f73cbba5f",
   "metadata": {},
   "source": [
    "### Simple Bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69ac9654-d189-4823-9550-a2618784270e",
   "metadata": {},
   "outputs": [],
   "source": [
    "imat_per_method(method=\"simple_bounds\", weights_dict=rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"expr_models\", \n",
    "                biomass_growth_df=biomass_growth_df)\n",
    "imat_per_method(method=\"simple_bounds\", weights_dict=norm_rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"norm_expr_models\", \n",
    "                biomass_growth_df=norm_biomass_growth_df)\n",
    "imat_per_method(method=\"simple_bounds\", weights_dict = diff_rxn_weight_dict, \n",
    "               results_path = BASE_PATH / \"results\" / \"diff_expr_models\",\n",
    "               biomass_growth_df = diff_biomass_growth_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c1fa61-84f1-4541-b956-c783d58aed02",
   "metadata": {},
   "source": [
    "### Subset Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82308696-6a41-4eb6-b9f5-b588bd36d4c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "imat_per_method(method=\"subset_model\", weights_dict=rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"expr_models\", \n",
    "                biomass_growth_df=biomass_growth_df)\n",
    "imat_per_method(method=\"subset_model\", weights_dict=norm_rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"norm_expr_models\", \n",
    "                biomass_growth_df=norm_biomass_growth_df)\n",
    "imat_per_method(method=\"subset_model\", weights_dict = diff_rxn_weight_dict, \n",
    "               results_path = BASE_PATH / \"results\" / \"diff_expr_models\",\n",
    "               biomass_growth_df = diff_biomass_growth_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1140e41-4ea9-47b5-8811-f5789ede4742",
   "metadata": {},
   "source": [
    "### FVA Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c6617e2-b20a-40ff-a610-1cae32afe042",
   "metadata": {},
   "outputs": [],
   "source": [
    "imat_per_method(method=\"fva_model\", weights_dict=rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"expr_models\", \n",
    "                biomass_growth_df=biomass_growth_df)\n",
    "imat_per_method(method=\"fva_model\", weights_dict=norm_rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"norm_expr_models\", \n",
    "                biomass_growth_df=norm_biomass_growth_df)\n",
    "imat_per_method(method=\"fva_model\", weights_dict = diff_rxn_weight_dict, \n",
    "               results_path = BASE_PATH / \"results\" / \"diff_expr_models\",\n",
    "               biomass_growth_df = diff_biomass_growth_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e915424-43cf-4c3c-a7c4-ee1b7e0b4575",
   "metadata": {},
   "source": [
    "### MILP Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b20e76-dec2-4c5a-be00-09674d5c2fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "imat_per_method(method=\"milp_model\", weights_dict=rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"expr_models\", \n",
    "                biomass_growth_df=biomass_growth_df)\n",
    "imat_per_method(method=\"milp_model\", weights_dict=norm_rxn_weight_dict, \n",
    "               results_path=BASE_PATH / \"results\" / \"norm_expr_models\", \n",
    "                biomass_growth_df=norm_biomass_growth_df)\n",
    "imat_per_method(method=\"milp_model\", weights_dict = diff_rxn_weight_dict, \n",
    "               results_path = BASE_PATH / \"results\" / \"diff_expr_models\",\n",
    "               biomass_growth_df = diff_biomass_growth_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4612dd6e-c044-4173-92d2-61c7b23aff14",
   "metadata": {},
   "source": [
    "### Save Biomass DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c21f9ee-d816-4957-9feb-83028202eaa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic results\n",
    "biomass_growth_df.to_csv(BASE_PATH / \"results\" / \"expr_models\" / \"biomass_growth.csv\")\n",
    "biomass_growth_df[~(biomass_growth_df.index==\"WT\")].div(biomass_growth_df.loc[\"WT\"], axis=1).to_csv(BASE_PATH / \"results\" / \"expr_models\" / \"biomass_growth_prop.csv\")\n",
    "# Qnorm results\n",
    "norm_biomass_growth_df.to_csv(BASE_PATH / \"results\" / \"norm_expr_models\" / \"norm_biomass_growth.csv\")\n",
    "norm_biomass_growth_df[~(norm_biomass_growth_df.index==\"WT\")].div(norm_biomass_growth_df.loc[\"WT\"], axis=1).to_csv(\n",
    "    BASE_PATH / \"results\" / \"norm_expr_models\" / \"norm_biomass_growth_prop.csv\")\n",
    "# Differential Expression Results\n",
    "diff_biomass_growth_df.to_csv(BASE_PATH / \"results\" / \"diff_expr_models\" / \"diff_biomass_growth.csv\")\n",
    "diff_biomass_growth_df[~(diff_biomass_growth_df.index==\"WT\")].div(diff_biomass_growth_df.loc[\"WT\"], axis=1).to_csv(\n",
    "    BASE_PATH / \"results\" / \"diff_expr_models\" / \"diff_biomass_growth_prop.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ae791a-3855-4b13-92f6-656a7701383a",
   "metadata": {},
   "source": [
    "## Essentiality Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a3ca9c3-5a38-4e3e-ab7c-0c4df13d1d3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def essentiality_analysis(model_dir: pathlib.Path, \n",
    "                          cond_list:list[str], \n",
    "                          wt:str,\n",
    "                          out_dir: pathlib.Path)->None: \n",
    "    wt_model = metworkpy.read_model(str(model_dir / f\"iek1011_{wt}_model.json\"))\n",
    "    flux_res_df = pd.DataFrame(np.NaN, \n",
    "                             dtype=\"float\", \n",
    "                             index = wt_model.genes.list_attr(\"id\"),\n",
    "                             columns = [wt]+cond_list\n",
    "                            )\n",
    "    prop_res_df = pd.DataFrame(np.NaN, \n",
    "                             dtype=\"float\", \n",
    "                             index = wt_model.genes.list_attr(\"id\"),\n",
    "                             columns = [wt]+cond_list\n",
    "                            )\n",
    "    ess_res_df = pd.DataFrame(np.NaN, \n",
    "                             dtype=\"float\", \n",
    "                             index = wt_model.genes.list_attr(\"id\"),\n",
    "                             columns = [wt]+cond_list\n",
    "                            )\n",
    "    for cond in [wt]+cond_list:\n",
    "        print(\"\\t-----------------\")\n",
    "        print(f\"\\tCondition: {cond}\")\n",
    "        print(\"\\t-----------------\")\n",
    "        cond_model = metworkpy.read_model(str(model_dir / f\"iek1011_{cond}_model.json\"))\n",
    "        try:\n",
    "            max_obj = cond_model.slim_optimize()\n",
    "            ko_res = single_gene_deletion(cond_model)\n",
    "            ko_res[\"ids\"] = ko_res[\"ids\"].apply(lambda x: next(iter(x)))\n",
    "            ko_res = ko_res.set_index(\"ids\")\n",
    "            flux_res_df.loc[ko_res.index, cond] = ko_res[\"growth\"]\n",
    "            prop_res_df.loc[flux_res_df.index, cond] = flux_res_df.loc[:, cond] / max_obj\n",
    "            ess_res_df.loc[(prop_res_df.loc[:,cond]<=5e-2).index, cond] = 1\n",
    "            ess_res_df.loc[(~(prop_res_df.loc[:,cond]>5e-2)).index, cond] = 0\n",
    "        except Exception as exp:\n",
    "            continue\n",
    "    out_dir.mkdir(parents=True, exist_ok = True)\n",
    "    flux_res_df.to_csv(out_dir / \"ko_flux.csv\")\n",
    "    prop_res_df.to_csv(out_dir / \"ko_prop.csv\")\n",
    "    ess_res_df.to_csv(out_dir / \"ko_essentiality.csv\")        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ea0935-f960-49d2-938d-14cccab4d03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normal Expression Essentiality Analysis\n",
    "EXPRESSION_METHOD = \"expr\"\n",
    "print(f\"Starting: {EXPRESSION_METHOD}\")\n",
    "for method in [\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"]:\n",
    "    essentiality_analysis(\n",
    "        model_dir = BASE_PATH / \"results\" / f\"{EXPRESSION_METHOD}_models\" / method,\n",
    "        cond_list = [\"PknB_KD\",\n",
    "                     \"PknB_IND\",\n",
    "                     \"PknD_KO\",\n",
    "                     \"PknD_IND\",\n",
    "                     \"PknE_KO\",\n",
    "                     \"PknE_IND\",\n",
    "                     \"PknF_KO\",\n",
    "                     \"PknF_IND\",\n",
    "                     \"PknG_KO\",\n",
    "                     \"PknG_IND\",\n",
    "                     \"PknH_KO\",\n",
    "                     \"PknH_IND\",\n",
    "                     \"PknJ_KO\",\n",
    "                     \"PknJ_IND\",\n",
    "                     \"PknK_KO\",\n",
    "                     \"PknK_IND\",\n",
    "                     \"PknL_KO\",\n",
    "                     \"PknL_IND\",],\n",
    "        wt = \"WT\",\n",
    "        out_dir = BASE_PATH / \"results\" / \"essentiality\" /  f\"{EXPRESSION_METHOD}_essentiality\" / method\n",
    "       )\n",
    "# Quantile Normalized Expression Essentiality Analysis\n",
    "EXPRESSION_METHOD = \"norm_expr\"\n",
    "print(f\"Starting: {EXPRESSION_METHOD}\")\n",
    "for method in [\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"]:\n",
    "    essentiality_analysis(\n",
    "        model_dir = BASE_PATH / \"results\" / f\"{EXPRESSION_METHOD}_models\" / method,\n",
    "        cond_list = [\"PknB_KD\",\n",
    "                     \"PknB_IND\",\n",
    "                     \"PknD_KO\",\n",
    "                     \"PknD_IND\",\n",
    "                     \"PknE_KO\",\n",
    "                     \"PknE_IND\",\n",
    "                     \"PknF_KO\",\n",
    "                     \"PknF_IND\",\n",
    "                     \"PknG_KO\",\n",
    "                     \"PknG_IND\",\n",
    "                     \"PknH_KO\",\n",
    "                     \"PknH_IND\",\n",
    "                     \"PknJ_KO\",\n",
    "                     \"PknJ_IND\",\n",
    "                     \"PknK_KO\",\n",
    "                     \"PknK_IND\",\n",
    "                     \"PknL_KO\",\n",
    "                     \"PknL_IND\",],\n",
    "        wt = \"WT\",\n",
    "        out_dir = BASE_PATH / \"results\" / \"essentiality\" /  f\"{EXPRESSION_METHOD}_essentiality\" / method\n",
    "       )\n",
    "# Differential Expression Essentiality Analysis\n",
    "EXPRESSION_METHOD = \"diff_expr\"\n",
    "print(f\"Starting: {EXPRESSION_METHOD}\")\n",
    "for method in [\"simple_bounds\", \"subset_model\", \"fva_model\", \"milp_model\"]:\n",
    "    essentiality_analysis(\n",
    "        model_dir = BASE_PATH / \"results\" / f\"{EXPRESSION_METHOD}_models\" / method,\n",
    "        cond_list = [\"PknB_KD\",\n",
    "                     \"PknB_IND\",\n",
    "                     \"PknD_KO\",\n",
    "                     \"PknD_IND\",\n",
    "                     \"PknE_KO\",\n",
    "                     \"PknE_IND\",\n",
    "                     \"PknF_KO\",\n",
    "                     \"PknF_IND\",\n",
    "                     \"PknG_KO\",\n",
    "                     \"PknG_IND\",\n",
    "                     \"PknH_KO\",\n",
    "                     \"PknH_IND\",\n",
    "                     \"PknJ_KO\",\n",
    "                     \"PknJ_IND\",\n",
    "                     \"PknK_KO\",\n",
    "                     \"PknK_IND\",\n",
    "                     \"PknL_KO\",\n",
    "                     \"PknL_IND\",],\n",
    "        wt = \"WT\",\n",
    "        out_dir = BASE_PATH / \"results\" / \"essentiality\" /  f\"{EXPRESSION_METHOD}_essentiality\" / method\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51bc129c-68b8-4ade-a163-1fdce4d3bb11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
